Lib/fontTools/otlLib/optimize.py

```diff
@@ -0,0 +1,487 @@
+    last, lastName, lastCls = items[0]
+    ranges = [[lastCls, last, lastName]]
+    for glyphID, glyphName, cls in items[1:]:
+        if glyphID != last + 1 or cls != lastCls:
+            ranges[-1].extend([last, lastName])
+            ranges.append([cls, glyphID, glyphName])
+        last = glyphID
+        lastName = glyphName
+        lastCls = cls
+    ranges[-1].extend([last, lastName])
+    return ranges
+
+
+# Adapted from https://github.com/fonttools/fonttools/blob/f64f0b42f2d1163b2d85194e0979def539f5dca3/Lib/fontTools/ttLib/tables/otTables.py#L960-L989
+def _classDef_bytes(font: TTFont, classes: Iterable[Tuple[str]]):
+    ranges = _getClassRanges(classes, font)
+    startGlyph = ranges[0][1]
+    endGlyph = ranges[-1][3]
+    glyphCount = endGlyph - startGlyph + 1
+    # https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#class-definition-table-format-1
+    format1_bytes = 6 + glyphCount * 2
+    # https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#class-definition-table-format-2
+    format2_bytes = 4 + len(ranges) * 6
+    return min(format1_bytes, format2_bytes)
+
+
+def cluster_pairs_by_class2_coverage_custom_cost(
+    font: TTFont,
+    pairs: Pairs,
+    compression: int = 5,
+) -> List[Pairs]:
+    # Sorted for reproducibility/determinism
+    all_class1 = sorted(set(pair[0] for pair in pairs))
+    all_class2 = sorted(set(pair[1] for pair in pairs))
+    # Use Python's big ints for binary vectors representing each line
+    lines = [
+        sum(
+            1 << i if (class1, class2) in pairs else 0
+            for i, class2 in enumerate(all_class2)
+        )
+        for class1 in all_class1
+    ]
+
+    # Precompute effective formats so that we can look them up within a cluster
+    # to compute the valueFormat1 and valueFormat2 of the whole cluster
+    # effective_value_formats: Dict[
+    #     Tuple[Tuple[str, ...], Tuple[str, ...]], Tuple[int, int]
+    # ] = defaultdict(lambda: (0, 0))
+    format1 = 0
+    format2 = 0
+    for pair, value in pairs.items():
+        # effective_value_formats[pair] = (
+        #     value[0].getEffectiveFormat() if value[0] else 0,
+        #     value[1].getEffectiveFormat() if value[1] else 0,
+        # )
+        format1 |= value[0].getEffectiveFormat() if value[0] else 0
+        format2 |= value[1].getEffectiveFormat() if value[1] else 0
+    valueFormat1_bytes = bit_count(format1) * 2
+    valueFormat2_bytes = bit_count(format2) * 2
+
+    # Agglomerative clustering by hand, checking the cost gain of the new
+    # cluster against the previously separate clusters
+    # Start with 1 cluster per line
+    # cluster = set of lines
+    # The class is here so it has a closure over lines and effective_value_formats
+    class Cluster:
+        # NOTE: (Jany) I wanted to use a dataclass it's Python 3.7+
+        # indices: int
+        # # Caches
+        # # NOTE: (Jany) I wanted to use functools.cached_property but it's Python 3.8+
```

Comment:
How about using `lru_cache`?

Reply:
The nice thing about @cached_property is that it saves typing but in the end I implemented the cached properties myself by using the _indices fields as the cache for the @property indices. I don't think lru_cache would help here. I could remove that comment actually since I have the caching in place?

Reply:
Leave the comment. Maybe we should start using a form like:
```
# TODO(Python 3.7): ...
```
such that when we bump up requirements, we can go clean such items.